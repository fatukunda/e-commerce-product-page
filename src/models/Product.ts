export interface IVariant {
  id: number;
  product_id: number;
  title: string;
  price: string;
  sku: string;
  position: number;
  compare_at_price: string;
  fulfillment_service: string;
  inventory_management: string;
  option1: string;
  option2?: string;
  option3?: string;
  created_at: string;
  updated_at: string;
  taxable: boolean;
  barcode: string;
  grams: number;
  image_id: number;
  weight: number;
  weight_unit: string;
  tax_code?: string;
  requires_shipping: boolean;
}

export interface IOption {
  id: number;
  product_id: number;
  name: string;
  position: number;
  values: string[];
}

export interface IImage {
  id: number;
  product_id: number;
  position: number;
  created_at: string;
  updated_at: string;
  alt?: string|null;
  width: number;
  height: number;
  src: string;
  variant_ids: number[];
}

export interface IProduct {
  id: number;
  title: string;
  body_html: string;
  vendor: string;
  product_type: string;
  created_at: string;
  handle: string;
  url: string;
  updated_at: string;
  published_at: string;
  variants: Array<
    {
      id: number;
      product_id: number;
      title: string;
      price: string;
      sku: string;
      position: number;
      compare_at_price: string;
      fulfillment_service: string;
      inventory_management: string;
      option1: string;
      option2?: string|null;
      option3?: string|null;
      created_at: string;
      updated_at: string;
      taxable: boolean;
      barcode: string;
      grams: number;
      image_id: number;
      weight: number;
      weight_unit: string;
      tax_code?: string;
      requires_shipping: boolean;
    }>
  options: [
    {
      id: number;
      product_id: number;
      name: string;
      position: number;
      values: string[];
    }
  ];
  images: Array<
    {
      id: number;
      product_id: number;
      position: number;
      created_at: string;
      updated_at: string;
      alt?: string|null;
      width: number;
      height: number;
      src: string;
      variant_ids: number[];
    }>;
  image: {
    id: number;
    product_id: number;
    position: number;
    created_at: string;
    updated_at: string;
    alt?: string|null;
    width: number;
    height: number;
    src: string;
    variant_ids: number[];
  };
}
