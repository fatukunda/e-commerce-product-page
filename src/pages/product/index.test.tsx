import { render, cleanup, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";
import { Provider as ReduxProvider } from "react-redux";

import Product from ".";
import mockStore from "src/components/mocks/store";

const renderProduct = () =>
  render(
    <ReduxProvider store={mockStore()}>
      <Product />
    </ReduxProvider>
  );

describe("Tests for <Product/>", () => {
  afterEach(cleanup);
  test("Product: Snapshot", () => {
    const component = renderer.create(
      <ReduxProvider store={mockStore()}>
        <Product />
      </ReduxProvider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("Renders without crashing", () => {
    renderProduct();
  });

  it("Selects a variant", () => {
    const { getAllByTestId, getByText } = renderProduct();
    const selectVariantBtns = getAllByTestId("select-variant-btn");
    fireEvent.click(selectVariantBtns[0]);
    expect(getByText("€18.99")).toBeInTheDocument();
  });

  it("increaments quantity of item", () => {
    const { getByTestId } = renderProduct();
    const increamentQuantity = getByTestId("increament-quantity-btn");
    const quantityText = getByTestId("quantity-text");
    expect(quantityText.textContent).toBe("1");
    fireEvent.click(increamentQuantity);
    expect(quantityText.textContent).toBe("2");
  });

  it("decreaments quantity of item", () => {
    const { getByTestId } = renderProduct();
    const decreamentQuantity = getByTestId("decreament-quantity-btn");
    const increamentQuantity = getByTestId("increament-quantity-btn");
    const quantityText = getByTestId("quantity-text");

    expect(quantityText.textContent).toBe("1");
    fireEvent.click(increamentQuantity);
    expect(quantityText.textContent).toBe("2");
    fireEvent.click(decreamentQuantity);
    expect(quantityText.textContent).toBe("1");
    fireEvent.click(decreamentQuantity);
    expect(quantityText.textContent).toBe("1");
  });

  it("adds to cart", () => {
    const { getByTestId, getByText, getAllByTestId } = renderProduct();
    const addToCart = getByTestId("main-btn");
    const selectVariantBtns = getAllByTestId("select-variant-btn");
    fireEvent.click(selectVariantBtns[0]);
    expect(getByText("€18.99")).toBeInTheDocument();
    fireEvent.click(addToCart);
  });
});
