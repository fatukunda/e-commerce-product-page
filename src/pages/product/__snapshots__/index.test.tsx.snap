// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <Product/> Product: Snapshot 1`] = `
<div
  className="px-6 py-4 pb-20 lg:max-w-lg"
>
  <span
    className="mb-4 inline-block text-xs font-bold uppercase tracking-widest text-primary lg:mb-6"
  >
    nu3
  </span>
  <h1
    className="mb-4 text-3xl font-bold text-tertiary md:text-4xl lg:mb-10 lg:text-5xl"
  >
    Test product
  </h1>
  <p
    className="mb-6 max-w-md text-tertiary-light"
  >
    This is a test product
  </p>
  <div
    className="space-y-4 md:flex md:items-center md:gap-4 md:space-y-0"
  >
    <div
      className="flex w-full items-center justify-between rounded-lg p-3 md:basis-6/12 mb-6 "
    >
      <button
        className="bg-white hover:bg-gray-100 text-gray-800 font-semibold py-2 px-4  border border-gray-400 rounded shadow"
        data-testid="select-variant-btn"
        onClick={[Function]}
      >
        150 g
      </button>
      <button
        className="bg-white hover:bg-gray-100 text-gray-800 font-semibold py-2 px-4  border border-gray-400 rounded shadow"
        data-testid="select-variant-btn"
        onClick={[Function]}
      >
        3 x 150 g
      </button>
    </div>
    <div>
      <div
        className="bg-gray-100 rounded-r-md flex flex-col items-center justify-center"
      >
        <div
          className="text-gray-500"
        >
          Your QrCode will show here...
        </div>
      </div>
    </div>
  </div>
  <div
    className="space-y-4 md:flex md:items-center md:gap-4 md:space-y-0"
  >
    <div
      className="flex w-full items-center justify-between rounded-lg bg-secondary-dark p-3 md:basis-8/12 "
    >
      <button
        className="p-3 transition-all duration-300 hover:opacity-50 lg:px-1"
        data-testid="decreament-quantity-btn"
        onClick={[Function]}
        type="button"
      >
        <img
          alt="minus"
          src="minus.svg"
        />
      </button>
      <span
        className="font-bold"
        data-testid="quantity-text"
      >
        1
      </span>
      <button
        className="p-3 transition-all duration-300 hover:opacity-50 lg:px-1"
        data-testid="increament-quantity-btn"
        onClick={[Function]}
        type="button"
      >
        <img
          alt="plus"
          src="plus.svg"
        />
      </button>
    </div>
    <button
      className="flex w-full items-center justify-center gap-2 rounded-lg bg-primary p-4 font-bold text-white shadow-lg shadow-primary/50 transition-all duration-300 hover:bg-primary/50"
      data-testid="main-btn"
      onClick={[Function]}
      type="button"
    >
      <svg
        className="scale-75 fill-white"
        height="22"
        width="21"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20.925 3.641H3.863L3.61.816A.896.896 0 0 0 2.717 0H.897a.896.896 0 1 0 0 1.792h1l1.031 11.483c.073.828.52 1.726 1.291 2.336C2.83 17.385 4.099 20 6.359 20c1.875 0 3.197-1.87 2.554-3.642h4.905c-.642 1.77.677 3.642 2.555 3.642a2.72 2.72 0 0 0 2.717-2.717 2.72 2.72 0 0 0-2.717-2.717H6.365c-.681 0-1.274-.41-1.53-1.009l14.321-.842a.896.896 0 0 0 .817-.677l1.821-7.283a.897.897 0 0 0-.87-1.114ZM6.358 18.208a.926.926 0 0 1 0-1.85.926.926 0 0 1 0 1.85Zm10.015 0a.926.926 0 0 1 0-1.85.926.926 0 0 1 0 1.85Zm2.021-7.243-13.8.81-.57-6.341h15.753l-1.383 5.53Z"
          fillRule="nonzero"
        />
      </svg>
      <span>
        Add to cart
      </span>
    </button>
  </div>
</div>
`;
